bugfix/refactor: camera bugfix/refactor + ninja invisible

kukurua opened the pull request:
## Описание
1. Motion Alarm в тревогах у ИИ и боргов вновь работают!
2. Так же была удалены area/ai_monitored из-за переработки зоны действия Motion Alarm. Зоны были заменены на такие же аналоги из уже существующих. Проблемы не замечены.
3. camera assembly теперь действительно может собирать камеры с улучшениями, а не просто держать в себе айтемы. 
4. Тревоги камер теперь работают через HasProximity.
5. По ниндзе в инвизе теперь не стреляют турели.
6. Тыкать ПДА или листок в камеру снова можно не только для ИИ, но и для тех кто смотрит камеру(да, это возможность просто не работала).

Вообще всё началось с баг фикса Motion Alarm, но потом пошло поехало и меня заставили делать рефактор камер(хоть и не большой). По мере отлаживания своих проблем заметил проблему с инвизом ниндзи и турелей и пофиксил.

## Ссылка на предложение/Причина создания ПР
Баг фикс Motion Alarm у ИИ и боргов.
https://discord.com/channels/617003227182792704/1238977359915388939

## Демонстрация изменений
Из-за фикса этого бага синты теперь будет знать, что кто-то проходит в поле зрения motion-sensitive security camera.

Commit:
Motion alarms for AI are working again.

Commit:
delete area ai_monitored

Commit:
Oops. Not the entire zone was removed. Now it shouldn't exist.

Commit:
"Camera assembly" now creates cameras that react to motion.

Commit:
А зачем тут ещё один if? А он тут и не нужен.

Daeberdir commented on code/game/area/areas.dm:
```diff
@@ -129,7 +131,16 @@
+				AddMotionCameraInList(M)
+
 	return INITIALIZE_HINT_LATELOAD
+/area/proc/AddMotionCameraInList(var/obj/machinery/camera/motion/M)
+	motioncameras.Add(M)
+	M.AddComponent(/datum/component/proximity_monitor)
+	M.set_area_motion(src)
```
Это выброси ниже, чтобы не лежало между `Initialize()` и `LateInitialize()` проками.

Daeberdir commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -23,6 +23,12 @@
 	QDEL_LIST(upgrades)
 	return ..()
 
+/obj/item/camera_assembly/proc/IsMotion(list/upgrades_list)
+	for(var/obj/I in upgrades_list)
+		if(I.type == /obj/item/assembly/prox_sensor)
+			return TRUE
+	return FALSE
```
```suggestion
/obj/item/camera_assembly/proc/IsMotion()
	for(var/obj/item/assembly/prox_sensor/I in upgrades)
		return TRUE
	return FALSE
```

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -69,7 +69,7 @@
 	if(isarea(myArea))
 		LAZYREMOVE(myArea.cameras, UID())
-	var/area/ai_monitored/A = get_area(src)
+	var/area/A = get_area(src)
 	if(istype(A))
 		A.motioncameras -= src
```
```suggestion
		A.motioncameras -= src
```
Но. Но может лучше увидеть тут лейзилист?

Daeberdir commented on code/game/area/areas.dm:
```diff
@@ -562,6 +578,14 @@
 	SEND_SIGNAL(src, COMSIG_AREA_EXITED, departed)
 	SEND_SIGNAL(departed, COMSIG_ATOM_EXITED_AREA, src)
 
+	if(!isliving(departed))
+		return
+
+	if(ismob(departed) && length(motioncameras)) //ai motion camera alarm deactivate
+		for(var/X in motioncameras)
+			var/obj/machinery/camera/cam = X
+			cam.lostTargetRef(departed.UID())
```
```suggestion
	if(!isliving(departed))
		return

	if(length(motioncameras)) //ai motion camera alarm deactivate
		for(var/obj/machinery/camera/cam as anything in motioncameras)
			cam.lostTargetRef(departed.UID())
```

Daeberdir commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -81,7 +87,13 @@
 	input = strip_html(input(usr, "How would you like to name the camera?", "Set Camera Name", temptag))
 	state = ASSEMBLY_BUILT
-	var/obj/machinery/camera/C = new(loc, uniquelist(tempnetwork))
+	var/obj/machinery/camera/C = null
+	if(IsMotion(upgrades)) //Check to create the right camera
+		C = new /obj/machinery/camera/motion(loc, uniquelist(tempnetwork))
+		camera_area.AddMotionCameraInList(C)
+	else
+		C = new /obj/machinery/camera(loc, uniquelist(tempnetwork))
+
```
`AddMotionCameraInList()` бы вытащить отсюда и из инициализации зон в инициализацию камеры. И ещё. Можно при создании камеры вот здесь делать только простую камеру, но в инициализацию передавать список апгрейдов, а не перереспавнивать эти апгрейды внутри камеры.

kukurua replied:
AddMotionCameraInList() перенёс в /obj/machinery/camera/proc/upgradeMotion(). Из area так же убран. Камеры создаю тут.

kukurua commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -81,7 +87,13 @@
+		C = new /obj/machinery/camera/motion(loc, uniquelist(tempnetwork))
+		camera_area.AddMotionCameraInList(C)
+	else
+		C = new /obj/machinery/camera(loc, uniquelist(tempnetwork))
```
Объявляю обычную камеру тут, а не выше, чтобы не создавать лишний раз камеру.

kukurua commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -23,6 +23,12 @@
+			return TRUE
+	return FALSE
+
 /obj/item/camera_assembly/attackby(obj/item/I, mob/living/user, params)
```
Если можно как-то по другому сверить содержимое листов по типу предметов, то буду рад подсказке.

Commit:
Now, in addition to a regular camera, you can create an anti-EMP camera, a motion-sensitive camera, a camera with Xray and a camera that includes all of the following.
AddMotionCameraInList moved to /obj/machinery/camera/proc/upgradeMotion().
Adding and removing a camera from the list of motioncameras now happens using LAZYADD and LAZYREMOVE

kukurua committed:
Merge branch 'master220' into ai_motion_area

kukurua commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -23,6 +23,26 @@
+		return TRUE
+	return FALSE
+
+/obj/item/camera_assembly/proc/isAll(list/upgrades_list)
```
Если есть варианты как лучше сверить, что список доступных апгрейдов == списку вставленных апгрейдов, то сообщите, пожалуйста.

kukurua commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -81,7 +101,17 @@
+	else if(isEmpProof(upgrades))
+		C = new /obj/machinery/camera/emp_proof(loc, uniquelist(tempnetwork))
+	else
+		C = new /obj/machinery/camera(loc, uniquelist(tempnetwork))
```
Объявляю обычную камеру тут чтобы не создавать два раза обычную камеру.

kukurua replied:
Считаю, что объявлять камеры лучше тут.

kukurua commented:
Кое что переделаю. Ожидайте нового комита.

Commit:
Окей. Теперь мы создаём обычную камеру, а уже в ней, при инициализации, проверяем какие в ней апгрейды и назначаем соответственные способности. Думаю это хорошее решение.

Daeberdir commented on code/game/area/areas.dm:
```diff
@@ -1,4 +1,6 @@
 /area
+	var/list/motioncameras = list()
+	var/list/motionTargets = list()
```
Окей. Эти листы передвинь ниже, к 65+ строчкам с другими листами, чтобы смотрелось лучше. Если действительно собираешься делать LAZYLIST, то лучше изучи специфику работы с ними. Они нужны, чтобы изначально было простое `var/list/moe_krutoe_nazvanie_lista` без пустого листа `= list()`, это если прям очень кратко.

Daeberdir commented on code/game/area/areas.dm:
```diff
@@ -543,6 +545,11 @@
+		for(var/X in motioncameras)
+			var/obj/machinery/camera/cam = X
+			cam.newTarget(arrived)
+
```
```suggestion
	if(LAZYLEN(motioncameras))	//ai motion camera alarm activate
		for(var/obj/machinery/camera/cam as anything in motioncameras)
			cam.newTarget(arrived)

```
Нету смысла проверять тип `arrived` на `/mob`, если мы на три строчки выше прекращаем выполнение функции при провале проверки типа на `/mob/living`.
И продолжение специфики работы с LAZYLIST.

Daeberdir commented on code/game/area/areas.dm:
```diff
@@ -562,6 +569,13 @@
+	if(ismob(departed) && length(motioncameras)) //ai motion camera alarm deactivate
+		for(var/obj/machinery/camera/cam as anything in motioncameras)
+			cam.lostTargetRef(departed.UID())
+
```
Коммент выше применяем и сюда.

Daeberdir commented on code/game/machinery/camera/presets.dm:
```diff
@@ -72,25 +73,22 @@
 	setPowerUsage()
 	// Add it to machines that process
 	START_PROCESSING(SSmachines, src)
+	src.myArea.AddMotionCameraInList(src)
```
```suggestion
	myArea.AddMotionCameraInList(src)
```
Не обязательно уточнять `src` перед проком или переменной, если они не были переопределены внутри функции.

Daeberdir commented on code/game/machinery/camera/presets.dm:
```diff
@@ -72,25 +73,22 @@
 
 /obj/machinery/camera/proc/upgradeXRay()
-	assembly.upgrades.Add(new /obj/item/analyzer(assembly))
+	icon_state = "xraycam"
```
Изменения `icon_state` нужно проводить через прок `update_icon_state()`, иначе будет некрасиво и **не** будет работать.

kukurua replied:
Но почему-то работает, хз

kukurua replied:
Но раз так лучше, то ок

Daeberdir replied:
Там сама суть в передаче сигнала о смене айкон стейта, который лежит внутри `update_icon_state()`. Тоже стоит изучить получше.

Daeberdir commented on code/game/machinery/camera/presets.dm:
```diff
@@ -72,25 +73,22 @@
 
 /obj/machinery/camera/proc/upgradeEmpProof()
-	assembly.upgrades.Add(new /obj/item/stack/sheet/mineral/plasma(assembly))
 	setPowerUsage()
```
`setPowerUsage()`  прок нету смысла прописывать нигде, кроме окончания инициализации камеры обычной.

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -35,14 +35,18 @@
 	var/toggle_sound = 'sound/items/wirecutter.ogg'
 
-/obj/machinery/camera/Initialize(mapload, list/networks)
+/obj/machinery/camera/Initialize(mapload, list/networks, list/upgrades, var/obj/item/camera_assembly/input_assembly)
```
```suggestion
/obj/machinery/camera/Initialize(mapload, list/networks, obj/item/camera_assembly/input_assembly)
```
Тут есть вариант создавать `camera_assembly` при инициализации пресетных камер, чтобы не ебаться с форсмувами и прочими радостями. Попробую расписать такой вариант... но лучше вынести список улучшений пресетных камер в отдельный список и спавнить их прямо тут, в инициализации обычной камеры.

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -35,14 +35,18 @@
-	assembly.set_anchored(TRUE)
-	assembly.update_icon(UPDATE_ICON_STATE)
-
+	if(input_assembly == null)
+		assembly = new(src)
+		assembly.state = 4
+		assembly.set_anchored(TRUE)
+		assembly.update_icon(UPDATE_ICON_STATE)
+		if(upgrades != null)
+			assembly.upgrades = upgrades
+	else
+		assembly = input_assembly
```
```suggestion
	if(!input_assembly)
		assembly = new(src)
		assembly.state = 4
		assembly.set_anchored(TRUE)
		assembly.update_icon(UPDATE_ICON_STATE)
	else
		assembly = input_assembly
```
По комментами ниже будет понятнее.

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -56,6 +60,13 @@
 	if(is_station_level(z) && prob(3) && !start_active)
 		toggle_cam(null, FALSE)
 		wires.cut_all()
+	for(var/obj/I in assembly.upgrades)
+		if(I.type == /obj/item/assembly/prox_sensor)
+			upgradeMotion()
+		if(I.type == /obj/item/stack/sheet/mineral/plasma)
+			upgradeEmpProof()
+		if(I.type == /obj/item/analyzer)
+			upgradeXRay()
```
Это можно сделать намного красивее, например, создав айтемам прок апгрейда камеры и уже внутри него прописав для этих трёх предметов их функции. А тут вызывать просто по такому типу:
```suggestion
	for(var/obj/item/I as anything in assembly.upgrades)
		I.upgrade_camera(src)
```

kukurua replied:
Нельзя так как я не хочу лезть ещё и в код малфа.

Daeberdir replied:
`upgrade_camera()` - просто условное название для прока, собственно, апгрейда камеры. Малф к этому ничего общего не имеет.

kukurua replied:
Не, там просто малф использует проки upgradeEmpProof() и upgradeXRay() в своих улучшениях.

kukurua replied:
То есть он не пересоздаёт камеры. А уже к существующим добавляет новые детали.
![изображение](https://github.com/ss220-space/Paradise/assets/39864130/4929cfa6-c1aa-406b-8270-1ed0ab139e3f)

kukurua replied:
Хотя, скорее всего можно будет переписать код на
```
for(var/V in GLOB.cameranet.cameras)
		var/obj/machinery/camera/C = V
		if(C.assembly)
			var/upgraded = FALSE
			if(!C.isXRay())
				C.assembly.upgrades.Add(new /obj/item/analyzer(C.assembly))
				for(/obj/item/analyzer/I in C.assembled.upgrades)
					I.camera_upgrade()
				//Update what it can see.
				upgraded = TRUE
			if(!C.isEmpProof())
				C.assembly.upgrades.Add(new new /obj/item/stack/sheet/mineral/plasma(C.assembly))
				for(/obj/item/stack/sheet/mineral/plasma/I in C.assembled.upgrades)
					I.camera_upgrade()
				upgraded = TRUE
			if(upgraded)
				upgraded_cameras++
```

Daeberdir replied:
![image](https://github.com/ss220-space/Paradise/assets/115735095/b114d311-aa7a-439c-82e2-2f3bee3ae670)
Как найдёшь приемлимое решение, либо забьёшь - позвать можно вот тут.

kukurua replied:
Вообщем, там для того чтобы for, который вы предложили, работал, надо в /obj/item сделать proc/camera_upgrade(var/obj/machinery/camera/input_cam) пустышку чтобы данный цикл работал(да и вообще компилился).
На сколько это будет правильное решение?

kukurua replied:
То есть proc в которм единственное действие - это return

Daeberdir replied:
Я бы так и сделал. И напоминаю о `/datum/AI_Module/large/upgrade_cameras` и `/obj/machinery/camera/attackby()`.

Daeberdir commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -81,10 +81,8 @@
 	input = strip_html(input(usr, "How would you like to name the camera?", "Set Camera Name", temptag))
 	state = ASSEMBLY_BUILT
-	var/obj/machinery/camera/C = new(loc, uniquelist(tempnetwork))
+	var/obj/machinery/camera/C = new /obj/machinery/camera(loc, uniquelist(tempnetwork),,src)
```
Если пытаешься вкидывать значения в проки вне последовательности, то либо вписывай последовательно стандартные значения:
`	var/obj/machinery/camera/C = new /obj/machinery/camera(loc, uniquelist(tempnetwork), upgrades, src)`
Либо определяй их в любой удобной последовательности:
`	var/obj/machinery/camera/C = new /obj/machinery/camera(loc, uniquelist(tempnetwork), input_assembly = src)`
Но продолжая тему с убиранием из функции передачи листа апгрейдов...
```suggestion
	var/obj/machinery/camera/C = new /obj/machinery/camera(loc, uniquelist(tempnetwork), src)
```

Daeberdir commented on code/game/machinery/camera/presets.dm:
```diff
@@ -3,35 +3,36 @@
-	. = ..()
-	upgradeEmpProof()
-
+	var/list/upgrades = list()
+	upgrades.Add(new /obj/item/stack/sheet/mineral/plasma)
+	. = ..(mapload,,upgrades)
```
```suggestion
	var/obj/item/camera_assembly/new_assembly = new(src)
	new_assembly.upgrades.Add(new /obj/item/stack/sheet/mineral/plasma(new_assembly))
	. = ..(input_assembly = new_assembly)
```
Вот такое должно работать. Но, конечно же, можно это всё дело сделать в инициализации простой камеры, добавляя туда апгрейды из отдельной переменной на камере.

Daeberdir commented on code/game/machinery/camera/presets.dm:
```diff
@@ -3,35 +3,36 @@
 // EMP
 
 /obj/machinery/camera/emp_proof/Initialize(mapload)
```
```suggestion
/obj/machinery/camera/emp_proof/Initialize(mapload, networks, input_assembly)
```
Вот это не забудь попрописывать во всех инициализациях пресетов камеры, если будешь оставлять эти инициализации.

kukurua replied:
Инициализации пресетов лучше оставлю, но вместо листов буду передавать camera_assembly

Daeberdir reviewed (commented):
Ну-с. Занимайся.

Commit:
Так как вариант с предложеным for оказался не рабочим, я создал отдельный метод camera_upgrade() в /obj/machenery/camera. Переписал немного способность малфа на улучшения камер иксрея и анти ЕМП.
Переделан файл presets. Теперь в камеру передаётся assembly, а не list апгрейдов.

kukurua committed:
Merge branch 'master220' into ai_motion_area

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -35,14 +35,16 @@
 	var/toggle_sound = 'sound/items/wirecutter.ogg'
 
-/obj/machinery/camera/Initialize(mapload, list/networks)
+/obj/machinery/camera/Initialize(mapload, list/networks, var/obj/item/camera_assembly/input_assembly)
```
```suggestion
/obj/machinery/camera/Initialize(mapload, list/networks, obj/item/camera_assembly/input_assembly)
```
Нету смысла уточнять, что это вар, внутри скобок функции.

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -35,14 +35,16 @@
 	. = ..()
 	wires = new(src)
-	assembly = new(src)
+	if(input_assembly == null)
+		assembly = new(src)
+	else
+		assembly = input_assembly
```
Если у тебя настолько протестантские чувства по отношению к знаку восклицания, то делай вот так:
```suggestion
	if(input_assembly)
		assembly = input_assembly
	else
		assembly = new(src)
```
Либо вот так:
```suggestion
	assembly = input_assembly ? input_assembly : new(src)
```

kukurua commented on code/game/machinery/camera/camera.dm:
```diff
@@ -99,6 +103,17 @@
 					to_chat(M, "The screen bursts into static.")
 			..()
 
+/obj/machinery/camera/proc/camera_upgrade()
+	if(isXRay(src) && !(icon_state == "xraycam" || icon_state == "xraycam1" || icon_state == "xraycamemp"))
+		update_icon(UPDATE_ICON_STATE)
+		//Update what it can see.
+		GLOB.cameranet.updateVisibility(src, 0)
+	if(isMotion(src) && name == initial(name))
+		name = "motion-sensitive security camera"
+		// Add it to machines that process
+		START_PROCESSING(SSmachines, src)
+		myArea.AddMotionCameraInList(src)
+	setPowerUsage()
```
Вместо того for придумал это. Мне кажется это удобнее.

kukurua replied:
проверка по иконке и по имени добавлена чтобы по несколько раз не добавлять способности камере и избежать повторного добавления в список. camera_upgrade() может вызываться несколько раз. Например: при использовании способности малфа или при добавлении уже рабочей камере новых компонентов.

kukurua commented on code/game/machinery/camera/camera.dm:
```diff
@@ -138,19 +153,33 @@
+		else
+			to_chat(user, "[msg2]")
+
+	else if(istype(I, /obj/item/analyzer) && panel_open)
+		if(!user.drop_transfer_item_to_loc(I, src))
+			to_chat(user, span_warning("[I] is stuck to your hand!"))
+			return
+		if(!isXRay())
+			assembly.upgrades.Add(I)
+			camera_upgrade()
+			update_icon(UPDATE_ICON_STATE)
 			add_fingerprint(user)
 			to_chat(user, "[msg]")
 			qdel(I)
```
В уже рабочую камеру можно добавить аналайзер для улучшения. Почему-то раньше нельзя было.

kukurua replied:
Хотя прокс сканер и плазму можно было. Забыли?

Daeberdir replied:
На то время не знали, как правильно менять иконку.

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -138,19 +153,33 @@
+			update_icon(UPDATE_ICON_STATE)
 			add_fingerprint(user)
 			to_chat(user, "[msg]")
 			qdel(I)
```
Тут и в двух случаях выше. Теперь апгрейды не доспавниваются внутри камер, ещё раз. Тебе нужно про установке не удалять их, а перемещать внутрь assembly камеры. Для плазмы пример есть в `/obj/machinery/constructable_frame/machine_frame/attackby()`

kukurua commented on code/game/gamemodes/malfunction/Malf_Modules.dm:
```diff
@@ -739,16 +739,15 @@
 
 			if(!C.isXRay())
-				C.upgradeXRay()
-				//Update what it can see.
-				GLOB.cameranet.updateVisibility(C, 0)
+				C.assembly.upgrades.Add(new /obj/item/analyzer(C.assembly))
 				upgraded = TRUE
 
 			if(!C.isEmpProof())
-				C.upgradeEmpProof()
+				C.assembly.upgrades.Add(new /obj/item/stack/sheet/mineral/plasma(C.assembly))
 				upgraded = TRUE
 
 			if(upgraded)
+				C.camera_upgrade()
 				upgraded_cameras++
```
Переделано немного с учётом camera_upgrade() и удалением старых проков.

kukurua commented on code/game/machinery/camera/camera.dm:
```diff
@@ -264,13 +293,28 @@
-		icon_state = "[initial(icon_state)]1"
-	else if(stat & EMPED)
-		icon_state = "[initial(icon_state)]emp"
+	if(assembly != null)
```
При разборке камеры assembly приравнивается к null, а isXray проверяет assembly.upgrades и соответственно выбивает ошибку из-за null.

Daeberdir replied:
Значит нужно решить проблему того, что оно обнуляет `assembly` раньше нужного.

Daeberdir replied:
Я показывал один из вариантов решения этой проблемы, просмотри ещё раз.

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -264,13 +293,28 @@
-		icon_state = "[initial(icon_state)]1"
-	else if(stat & EMPED)
-		icon_state = "[initial(icon_state)]emp"
+	if(assembly != null)
```
`icon_state = isXRay(src) ? "xraycam" : initial(icon_state)`
`	if(!status)`
`		icon_state = "[icon_state]1"`
`	else if(stat & EMPED)`
`		icon_state = "[icon_state]emp"`
Это тоже можно сделать лучше, но нужно будет залезать в файл со спрайтами.

Daeberdir commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -81,10 +81,8 @@
 	input = strip_html(input(usr, "How would you like to name the camera?", "Set Camera Name", temptag))
 	state = ASSEMBLY_BUILT
-	var/obj/machinery/camera/C = new(loc, uniquelist(tempnetwork))
+	var/obj/machinery/camera/C = new /obj/machinery/camera(loc, uniquelist(tempnetwork),src)
```
```suggestion
	var/obj/machinery/camera/C = new(loc, uniquelist(tempnetwork), src)
```
Оно заспавнит обьект С таким, каким об был определён.

Daeberdir commented on code/game/machinery/camera/presets.dm:
```diff
@@ -1,38 +1,37 @@
 
+/obj/machinery/camera/all/Initialize(mapload, networks, input_assembly)
+	var/obj/item/camera_assembly/new_assembly = new(src)
+	new_assembly.upgrades.Add(new /obj/item/stack/sheet/mineral/plasma(new_assembly))
+	new_assembly.upgrades.Add(new /obj/item/assembly/prox_sensor(new_assembly))
+	new_assembly.upgrades.Add(new /obj/item/analyzer(new_assembly))
```
```suggestion
	new_assembly.upgrades.Add(new /obj/item/stack/sheet/mineral/plasma(new_assembly),
				new /obj/item/assembly/prox_sensor(new_assembly),
				new /obj/item/analyzer(new_assembly))
```
`Add()` поддерживает использование нескольких обектов сразу.

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -99,6 +103,17 @@
 					to_chat(M, "The screen bursts into static.")
 			..()
 
+/obj/machinery/camera/proc/camera_upgrade()
+	if(isXRay(src) && !(icon_state == "xraycam" || icon_state == "xraycam1" || icon_state == "xraycamemp"))
+		update_icon(UPDATE_ICON_STATE)
+		//Update what it can see.
+		GLOB.cameranet.updateVisibility(src, 0)
+	if(isMotion(src) && name == initial(name))
+		name = "motion-sensitive security camera"
+		// Add it to machines that process
+		START_PROCESSING(SSmachines, src)
+		myArea.AddMotionCameraInList(src)
+	setPowerUsage()
```
Вот эта часть плохо выглядит. Проще будет завтра в дисе обьяснить...

Commit:
We add plasma correctly and make small adjustments.

kukurua committed:
Merge branch 'master220' into ai_motion_area

Commit:
Merge branch 'ai_motion_area' of https://github.com/kukurua/Paradise into ai_motion_area

kukurua committed:
Merge branch 'ss220-space:master220' into ai_motion_area

Commit:
Merge branch 'ai_motion_area' of https://github.com/kukurua/Paradise into ai_motion_area

Commit:
The code for adding upgrades to an already working camera has been changed + minor changes.

kukurua committed:
Merge branch 'master220' into ai_motion_area

Daeberdir commented on code/game/area/areas.dm:
```diff
@@ -570,6 +583,10 @@
 		if(!prev_gravity && user.gravity_state)
 			user.thunk()
 
+/area/proc/AddMotionCameraInList(var/obj/machinery/camera/M)
```
```suggestion
/area/proc/AddMotionCameraInList(obj/machinery/camera/M)
```

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -98,7 +103,20 @@
+// apply the improvements inserted into the camera
+// Use this if you have added improvements to assembly.upgrades
+/obj/machinery/camera/proc/camera_upgrade()
+	if(isXRay(src) && !isAlreadyUpgradeXray)
```
```suggestion
	if(isXRay() && !isAlreadyUpgradeXray)
```
Ну... я не знаю, стоит ли нагнетать третий раз...

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -98,7 +103,20 @@
+		update_icon(UPDATE_ICON_STATE)
+		//Update what it can see.
+		GLOB.cameranet.updateVisibility(src, 0)
+	if(isMotion(src) && name == initial(name))
```
```suggestion
	if(isMotion() && name == initial(name))
```

Daeberdir commented on code/game/machinery/camera/camera.dm:
```diff
@@ -407,7 +427,7 @@
 
 /obj/machinery/camera/update_remote_sight(mob/living/user)
-	if(isXRay() && isAI(user))
+	if(isAlreadyUpgradeXray && isAI(user))
```
На этом месте всё стало выглядеть ещё хуже...

Daeberdir commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -81,10 +92,8 @@
 	input = strip_html(input(usr, "How would you like to name the camera?", "Set Camera Name", temptag))
 	state = ASSEMBLY_BUILT
-	var/obj/machinery/camera/C = new(loc, uniquelist(tempnetwork))
+	var/obj/machinery/camera/C = new(loc, uniquelist(tempnetwork),src)
```
```suggestion
	var/obj/machinery/camera/C = new(loc, uniquelist(tempnetwork), src)
```
Чи-та-бель-ность.

Daeberdir reviewed (commented):
Добавление переменной `isAlreadyUpgradeXray` выглядит плохо. Завтра могу показать ещё один вариант решения...

kukurua committed:
Update code/game/machinery/camera/camera.dm

Co-authored-by: Daeberdir <115735095+Daeberdir@users.noreply.github.com>

kukurua committed:
Update code/game/area/areas.dm

Co-authored-by: Daeberdir <115735095+Daeberdir@users.noreply.github.com>

kukurua committed:
Update code/game/machinery/camera/camera.dm

Co-authored-by: Daeberdir <115735095+Daeberdir@users.noreply.github.com>

kukurua committed:
Update code/game/machinery/camera/camera_assembly.dm

Co-authored-by: Daeberdir <115735095+Daeberdir@users.noreply.github.com>

Commit:
Porting Daeberdir improvements

kukurua committed:
Merge branch 'master220' into ai_motion_area

Commit:
Minor improvements to the logic and readability of the code.

kukurua committed:
Merge branch 'master220' into ai_motion_area

Commit:
Minor improvements

Commit:
Minor edits and changes

kukurua committed:
Merge branch 'master220' into ai_motion_area

Commit:
Bruh

Commit:
Merge branch 'ai_motion_area' of https://github.com/kukurua/Paradise into ai_motion_area

Commit:
Now cameras that respond to movements work not through “entered the area”, but through HasProximity. The bug that turrets shoot at ninjas who are in invisibility has been fixed.

Commit:
Removing unnecessary and bug fixes

Commit:
The motion alarm is now triggered if the camera sees a mob and it is in the same area as the camera.

Commit:
The next small improvements

Commit:
small improvements

kukurua committed:
Delete .vscode/launch.json

Commit:
small improvements 2

kukurua committed:
Merge branch 'master220' into ai_motion_area

Commit:
Small changes 3

Daeberdir commented:
Есть куда расти. Но приемлимо.

Gottfrei commented on code/game/machinery/camera/camera.dm:
```diff
@@ -240,6 +231,33 @@
+	START_PROCESSING(SSmachines, target)
+	target.AddComponent(/datum/component/proximity_monitor, target.view_range, TRUE)
+
+/obj/machinery/camera/update_name(updates)
+	. = ..()
+	if(isMotion())
+		name = "motion-sensitive security camera"
```
дефолтное имя всегда должно быть как альтернатива.

Gottfrei commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -36,6 +36,17 @@
+				to_chat(user, span_warning("[I] is stuck!"))
+				return
+			var/obj/item/stack/sheet/mineral/plasma/new_stack = new(src, 1)
+			new_stack.update_icon(UPDATE_ICON_STATE)
```
то же самое по поводу апдейта иконок.

Gottfrei commented on code/game/machinery/camera/camera.dm:
```diff
@@ -133,29 +129,24 @@
+				to_chat(user, span_warning("[I] is stuck to your hand!"))
+				return
+			var/obj/item/stack/sheet/mineral/plasma/upgrade = new(assembly, 1)
+			upgrade.update_icon(UPDATE_ICON_STATE)
```
нам не нужно апдейтить иконки у стака.

вообще лучше всего проверять стак вот так:
```dm
if(!use(1))
    return
```

Gottfrei commented on code/game/machinery/camera/motion.dm:
```diff
@@ -58,9 +56,27 @@
+	var/turf/target_turf = get_turf(target)
+	if(target.invisibility > SEE_INVISIBLE_LIVING || target.alpha == NINJA_ALPHA_INVISIBILITY)
+		return FALSE
+	if(current != target_turf)
+		if(get_dist(current, target_turf) > view_range)
+			return FALSE
+		if(!isXRay())
+			while(current != target_turf)
+				current = get_step_towards(current, target_turf)
+				if(current.opacity)
+					return FALSE
+				for(var/thing in current)
+					var/atom/A = thing
+					if(A.opacity)
+						return FALSE
+	return TRUE
```
перепиши без лесенок, с ранними ретурнами.

Gottfrei commented on code/game/machinery/camera/motion.dm:
```diff
@@ -58,9 +56,27 @@
+			return FALSE
+		if(!isXRay())
+			while(current != target_turf)
+				current = get_step_towards(current, target_turf)
```
турфа может не быть, если мы упрёмся в границы.

Commit:
Changing can_see in motion and minor improvements 4

Commit:
Minor changes 5

Commit:
We remove what is not in use and fix the paper sheet for the security consoles.

Gottfrei commented on code/game/machinery/camera/camera_assembly.dm:
```diff
@@ -36,6 +36,16 @@
 	// Upgrades!
 	else if(is_type_in_list(I, possible_upgrades) && !is_type_in_list(I, upgrades)) // Is a possible upgrade and isn't in the camera already.
+		if(isstack(I))
+			if(!user.can_unEquip(I))
+				to_chat(user, span_warning("[I] is stuck!"))
+				return
+			var/obj/item/stack/sheet/mineral/plasma/new_stack = new(src, 1)
+			to_chat(user, span_notice("You attach [I] into the assembly inner circuits."))
+			upgrades += new_stack
+			I.use(1)
```
проверка на use.

Gottfrei commented on code/game/machinery/camera/motion.dm:
```diff
@@ -58,9 +56,30 @@
+	if(current == target_turf || isXRay())
+		return TRUE
+
+	var/list/line_of_sight = get_line(src, target)
+	for(var/turf/current_turf as anything in line_of_sight)
+		if(!current_turf)
+			return FALSE
+		if(current_turf.opacity)
+			return FALSE
+		for(var/thing as anything in current_turf)
+			var/atom/A = thing
+			if(A.opacity)
+				return FALSE
+	return TRUE
```
```suggestion
	var/list/line_of_sight = get_line(src, target)
	line_of_sight = line_of_sight.Cut(1, 2)
	for(var/turf/current_turf as anything in line_of_sight)
		if(current_turf.opacity)
			return FALSE
		for(var/atom/movable/thing as anything in current_turf)
			if(thing.opacity)
				return FALSE
	return TRUE
```
нам нужно резать первый турф line_of_sight.Cut(1, 2) мы уже проверяли на него выше.

Commit:
Minor changes 6

Daeberdir reviewed (dismissed):
![cat-applause](https://github.com/ss220-space/Paradise/assets/115735095/91244088-b382-499c-b9ce-04598c163635)

kukurua commented:
Уээээ. Мерж конфликт ещё и в мапе.

Daeberdir commented:
Каво...
![изображение](https://github.com/ss220-space/Paradise/assets/115735095/dbeb41af-29dc-45e0-a4b5-eaca27f45ab7)

kukurua commented:
Я обязательно пойму как это починить....

kukurua commented:
Просто беру отсюда карту(https://github.com/ss220-space/Paradise/pull/5145), вставляю себе, редактирую немного и она не проходит проверку... Мапа компилится у меня...

Commit:
Удалим мапу

Commit:
Переносим новую мапу с изменениями

kukurua commented:
Удалил у себя коробку, закоммитил, перетащил коробку с этого коммита https://github.com/ss220-space/Paradise/pull/5145 и отредактировал зоны, которые у меня были удалены. Поэтому +80 000 строк кода.

Den109G commented:
![image](https://github.com/ss220-space/Paradise/assets/87372121/138d9031-7e9b-4a52-983f-06ddd1253445)
Ну, что-то много изменилось на коробке.

kukurua commented:
Я пожалуй просто новую ветку на основе мастера создам и перенесу изменения из текущей не включая карты. Карты в ручную уже поправлю. ^_^
